internal/subshell/retry_test.go

```diff
@@ -0,0 +1,343 @@
+		scriptPath := filepath.Join(tmpDir, "count-attempts.sh")
+		scriptContent := `#!/bin/bash
+COUNTER_FILE="` + counterFile + `"
+if [ ! -f "$COUNTER_FILE" ]; then
+    echo "1" > "$COUNTER_FILE"
+else
+    COUNT=$(cat "$COUNTER_FILE")
+    echo $((COUNT + 1)) > "$COUNTER_FILE"
+fi
+>&2 echo "fatal: Unable to create '.git/index.lock': File exists."
+exit 1
+`
+		must.NoError(t, os.WriteFile(scriptPath, []byte(scriptContent), 0o755))
+
+		_, err := runner.Query("bash", scriptPath)
+		must.Error(t, err)
+
+		// Read the counter to verify it was called exactly 5 times
+		counterBytes, err := os.ReadFile(counterFile)
+		must.NoError(t, err)
+		must.EqOp(t, "5\n", string(counterBytes))
+	})
+
+	t.Run("does not retry on non-lock errors", func(t *testing.T) {
+		t.Parallel()
+		tmpDir := t.TempDir()
+		runner := subshell.BackendRunner{Dir: Some(tmpDir), Verbose: false, CommandsCounter: NewMutable(new(gohacks.Counter))}
+
+		// Create a script that fails with a different error
+		scriptPath := filepath.Join(tmpDir, "other-error.sh")
+		scriptContent := `#!/bin/bash
+>&2 echo "fatal: Some other error"
+exit 1
+`
+		must.NoError(t, os.WriteFile(scriptPath, []byte(scriptContent), 0o755))
+
+		start := time.Now()
+		_, err := runner.Query("bash", scriptPath)
+		duration := time.Since(start)
+
+		// Should fail immediately
+		must.Error(t, err)
+		// Should complete quickly without retries
+		must.Less(t, 500*time.Millisecond, duration)
+	})
+
+	t.Run("detects index.lock error pattern correctly", func(t *testing.T) {
+		t.Parallel()
+		tmpDir := t.TempDir()
+		runner := subshell.BackendRunner{Dir: Some(tmpDir), Verbose: false, CommandsCounter: NewMutable(new(gohacks.Counter))}
+
+		testCases := []struct {
+			name        string
+			errorMsg    string
+			shouldRetry bool
+		}{
+			{
+				name:        "exact match",
+				errorMsg:    "fatal: Unable to create '.git/index.lock': File exists.",
+				shouldRetry: true,
+			},
+			{
+				name:        "with path variations",
+				errorMsg:    "fatal: Unable to create '/path/to/repo/.git/index.lock': File exists.",
+				shouldRetry: true,
+			},
+			{
+				name:        "missing 'fatal' prefix but has both patterns",
+				errorMsg:    "Unable to create '.git/index.lock': File exists.",
+				shouldRetry: false, // doesn't match because "fatal: Unable to create" is required
+			},
+			{
+				name:        "only has 'Unable to create'",
+				errorMsg:    "fatal: Unable to create '.git/something'",
+				shouldRetry: false,
+			},
+			{
+				name:        "only has 'File exists'",
+				errorMsg:    "index.lock': File exists.",
+				shouldRetry: false,
+			},
+			{
+				name:        "completely different error",
+				errorMsg:    "fatal: not a git repository",
+				shouldRetry: false,
+			},
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				scriptPath := filepath.Join(tmpDir, fmt.Sprintf("test-%s.sh", tc.name))
+				scriptContent := fmt.Sprintf(`#!/bin/bash
+>&2 echo "%s"
```

Comment:
You can let Go quote the string for you.

```suggestion
>&2 echo %q
```